cmd/openshift-compatibility-gen/generators/gen_type.go

```diff
@@ -0,0 +1,197 @@
+		return false
+	}
+	return true
+}
+
+func (g *genCompatibility) Init(c *generator.Context, w io.Writer) error {
+	return nil
+}
+
+func (g *genCompatibility) GenerateType(c *generator.Context, t *types.Type, w io.Writer) error {
+	klog.V(3).Infof("Generating OpenShift compatibility for type %v", t)
+
+	sw := generator.NewSnippetWriter(w, c, "$", "$")
+	args, err := g.argsFromType(c, t)
+	if err != nil {
+		return err
+	}
+
+	if versionedMethodOrDie("OpenShiftAPICompatibilityLevel", t) == nil {
+		sw.Do("// OpenShiftAPICompatibilityLevel is an autogenerated function, returning the openshift api compatibility level.\n", args)
+		sw.Do("// It is controlled by \""+levelTagName+"\" tags in types.go.\n", args)
+		sw.Do("// Compatibility level $.level$: $.comment$\n", args)
+		sw.Do("func (in *$.type|intrapackage$) OpenShiftAPICompatibilityLevel() (level int) {\n", args)
+		sw.Do("    return $.level$\n", args)
+		sw.Do("}\n\n", nil)
+	}
+	return sw.Error()
+}
+
+const (
+	level1Comment = "Stable within a major release for a minimum of 12 months or 3 minor releases (whichever is longer)."
+	level2Comment = "Stable within a major release for a minimum of 9 months or 3 minor releases (whichever is longer)."
+	level3Comment = "Will attempt to be as compatible from version to version as possible, but version to version compatibility is not guaranteed"
+	level4Comment = "No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support."
+)
+
+func (g *genCompatibility) argsFromType(c *generator.Context, t *types.Type) (generator.Args, error) {
+	a := generator.Args{
+		"type": t,
+	}
+
+	l, ok := extractOpenShiftCompatibilityLevelTag(t)
+	if !ok {
+		return a, fmt.Errorf("%s: level must be specified", t.Name)
+	}
+
+	switch {
+	case versionIsGenerallyAvailable(t):
+		if l != 1 {
+			return a, fmt.Errorf("%s: generally available APIs must be supported for a minimum of 12 months", t.Name)
+		}
+	case versionIsPrerelease(t):
+		switch l {
+		case 1:
+			return a, fmt.Errorf("%s: pre-release (beta) APIs must offer level 2 compatibility: the %s API should be versioned as generally available if you with to offer level 1 compatibility", t.Name, t.Name.Name)
+		case 4:
+			return a, fmt.Errorf("%s: pre-release (beta) APIs must offer level 2 compatibility: the %s API should be versioned as experimental (alpha) if you wish to offer level 4 compatibility", t.Name, t.Name.Name)
+		}
+	case versionIsExperimental(t):
+		if l != 4 {
+			return a, fmt.Errorf("%s: experimental (alpha) APIs are only allowed to offer level 4 compatibility: long term support cannot be offered for the %s API", t.Name, t.Name.Name)
+		}
+	default:
+		return a, fmt.Errorf("%s: the %s package name does not conform to the API version package naming conventions", t.Name, path.Base(path.Dir(t.Name.Package)))
+	}
+
+	a = a.With("level", l)
+	switch l {
+	case 1:
+		a = a.With("comment", level1Comment)
```

Comment:
what does thsi do?

Reply:
> what does thsi do?

Oh, I see.  That's fine, but in a followup we want to force it onto the doc for the type itself so it shows up in openapi.